### ISSUE TITLE
Status Code: 404

### ISSUE THREAD

**Issue by DestyKarpov** [gigobyte/HLTV#59]
Greetings! Used your development, everything worked perfectly! But recently I began to notice that instead of matches began to come an empty array. I decided to check. I sent a request on the link https://www.hltv.org/matches with the REQUEST module. The status code is 404, but the content of the page is successfully taken. Perhaps the administration of the project hltv.org made such a "bluff" to the parsers stopped working. I ask for help!

**Comment by twanh** [comment]
Perhaps it has something to do with the fact that cloud scraper automatically returns a request if its 404. I am not sure of this though.

But since the content still is present on the page I suspect that the 404 error causes the parser/request(cloudscaper) to break. 

So if this is the case (which I am again not sure of, I am not really familiar with the code base and unable to check it out currently) you could possibly fix this by making sure the 404 error is ignored and the parsing continues as it should.

TLDR; Check if there is some exception handling 404 errors by default (that could explain the empty array whilst still having the actual content pressent on the page). If so you could try to change the exception so that it ignores the 404 and continues parsing as if there were no 404.
Btw if you find the solution and it required chante in the hltv code base make a pull requested :).

I hope it helpts! I will check this out myself when I get home in 3weeks :/

**Comment by twanh** [comment]
hltv.com has added CAPTCHA I expect this has to do with that?
Check issue #57 for that.

**Comment by gigobyte** [comment]
> The status code is 404, but the content of the page is successfully taken. 
Can you confirm this? It returns a page with a CAPTCHA when I run it.

**Comment by twanh** [comment]
Perhaps that has to do with the frequency of the requests of ip address? Usually CAPTCHAs come in to play when a request is suspected to be from a bot, factors in that might be ip address, frequency, missing headers etc...

But if the request does return a page with error 404 with all the contents present the problem (other then the newly added CAPTCHA) might be that the parser/request breaks or has an exception because of the 404 and therefore does not attempt to extract the contents from the page. If this is the case the fix might be to change how 404 errors are handled...?

**Comment by gigobyte** [comment]
I don't think it's per IP because I haven't launched the library in about a month and I only use it during development.

**Comment by twanh** [comment]
Yeah true. Now that I think about it, you probably would have some trouble when going to the site using your regular browser I guess because it's still the same ip.

So because of this CAPTCHA wouldn't the whole library stop working because requests cannot get through anymore?

**Comment by gigobyte** [comment]
You can access the site through the browser normally, but cloudflare can detect bots.

I tried other hltv libraries and they don't seem to work too, seems like this library will have to stay on hold until HLTV removes the protection or until somebody can solve the problem.

**Comment by twanh** [comment]
Yep. There is no solution for bypassing CAPTCHA that I know of as of now. I will do some more extensive research though.

Perhaps should this problem get it's own issue?

**Comment by gigobyte** [comment]
I'll keep this one open because when I closed the previous ones people just kept making new ones :sweat_smile:

**Comment by DestyKarpov** [comment]
Now I decided to study the parser code and make my own adjustments. In the file /utils/mappers.js instead of the library **cloudScraper** connected the **request** and everything worked!

![image](https://user-images.githubusercontent.com/24729953/44337497-7be98880-a4ad-11e8-83c5-933819e371d2.png)

After all the changes, I ran the parser (**getMatches** method) and got this response:

`[{"id":2325735,"team1":{"name":"MVP PK","id":7354},"team2":{"name":"CyberZen","id":6503},"event":{"name":"StarSeries i-League Season 6 Asia Qualifier","id":3937},"format":"Best of 3","maps":["inf","nuke","mrg"],"stars":1,"live":true},{"id":2325829,"team1":{"name":"AGO","id":8068},"team2":{"name":"3DMAX","id":4914},"event":{"name":"EPICENTER 2018 Europe Open Qualifier #1","id":3992},"format":"Best of 3","maps":["cch","inf","d2"],"stars":1,"live":true},{"id":2325830,"team1":{"name":"Royal Bandits","id":8263},"team2":{"name":"Smoke Criminals","id":8369},"event":{"name":"EPICENTER 2018 Europe Open Qualifier #1","id":3992},"format":"Best of 3","maps":["nuke","cch","inf"],"stars":0,"live":true},{"id":2325835,"date":1534766400000,"team1":{"name":"Flash","id":7660},"team2":{"name":"MVP PK","id":7354},"format":"bo3","event":{"name":"IEM Chicago 2018 Asia Closed Qualifier","id":3970},"stars":1,"live":false},{"id":2325836,"date":1534766400000,"team1":{"name":"SCARZ Absolute","id":7962},"team2":{"name":"5POWER","id":7092},"format":"bo3","event":{"name":"IEM Chicago 2018 Asia Closed Qualifier","id":3970},"stars":0,"live":false},{"id":2325837,"date":1534766400000,"team1":{"name":"MiTH","id":6353},"team2":{"name":"BOOT-d[S]","id":7567},"format":"bo3","event":{"name":"IEM Chicago 2018 Asia Closed Qualifier","id":3970},"stars":0,"live":false},{"id":2325831,"date":1534770000000,"title":"Grand final - EPICENTER 2018 EU Open Qual #1","stars":0,"live":false},{"id":2325508,"date":1534773600000,"team1":{"name":"fnatic Academy","id":7169},"team2":{"name":"Bpro","id":6762},"format":"bo3","event":{"name":"Thunderpick Invitational #3","id":3946},"stars":0,"live":false},{"id":2325781,"date":1534773600000,"team1":{"name":"Chaos","id":9085},"team2":{"name":"SuperJymy","id":8504},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325783,"date":1534777200000,"team1":{"name":"PACT","id":8248},"team2":{"name":"Tricked","id":4602},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325834,"date":1534777200000,"team1":{"name":"EURONICS","id":6343},"team2":{"name":"Sharks","id":8113},"format":"bo3","event":{"name":"GG.BET Summer Europe","id":3960},"stars":1,"live":false},{"id":2325608,"date":1534798800000,"team1":{"name":"PITS","id":9265},"team2":{"name":"Bulldozer","id":9013},"format":"bo3","event":{"name":"GG.BET Summer Brazil","id":3955},"stars":0,"live":false},{"id":2325609,"date":1534806000000,"team1":{"name":"Wild","id":8895},"team2":{"name":"FURIA Inagame","id":8885},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325610,"date":1534810200000,"team1":{"name":"DETONA","id":8902},"team2":{"name":"C4","id":8051},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325842,"date":1534813200000,"team1":{"name":"DETONA","id":8902},"team2":{"name":"ex-Wild","id":9370},"format":"bo5","event":{"name":"GG.BET Summer Brazil","id":3955},"stars":0,"live":false},{"id":2325611,"date":1534814400000,"team1":{"name":"Coscu Army","id":8871},"team2":{"name":"Furious","id":6799},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325650,"date":1534815000000,"team1":{"name":"Mentality","id":9187},"team2":{"name":"Reform","id":9279},"format":"bo3","event":{"name":"ESEA Advanced Season 28 North America","id":3909},"stars":0,"live":false},{"id":2325612,"date":1534818600000,"team1":{"name":"YeaH!","id":8728},"team2":{"name":"Isurus","id":7653},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325838,"date":1534831200000,"title":"IEM Chicago Asia Qualifier - Lower bracket final","stars":0,"live":false},{"id":2325839,"date":1534842000000,"title":"IEM Chicago Asia Qualifier - Consolidation Final","stars":0,"live":false},{"id":2325841,"date":1534852800000,"title":"IEM Chicago Asia Qualifier - Grand Final","stars":0,"live":false},{"id":2325792,"date":1534856400000,"team1":{"name":"WINside","id":9055},"team2":{"name":"Unique","id":9287},"format":"bo3","event":{"name":"Stream.me CIS Challenger League #2","id":3940},"stars":0,"live":false},{"id":2325649,"date":1534870800000,"team1":{"name":"Endpoint","id":7234},"team2":{"name":"Squared","id":7669},"format":"bo3","event":{"name":"Thunderpick Invitational #3","id":3946},"stars":0,"live":false},{"id":2325780,"date":1534878000000,"team1":{"name":"Nexus","id":7187},"team2":{"name":"x-kom","id":8846},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325499,"date":1534881600000,"team1":{"name":"W7M","id":8574},"team2":{"name":"INTZ","id":6619},"format":"bo3","event":{"name":"Brasil Game Cup 2018","id":3814},"stars":0,"live":false},{"id":2325613,"date":1534892400000,"team1":{"name":"Wild","id":8895},"team2":{"name":"Furious","id":6799},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325614,"date":1534896600000,"team1":{"name":"C4","id":8051},"team2":{"name":"Isurus","id":7653},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325615,"date":1534900800000,"team1":{"name":"Coscu Army","id":8871},"team2":{"name":"FURIA Inagame","id":8885},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325616,"date":1534905000000,"team1":{"name":"YeaH!","id":8728},"team2":{"name":"DETONA","id":8902},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325784,"date":1534946400000,"team1":{"name":"GoodJob","id":7820},"team2":{"name":"FriendlyFire","id":9062},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325507,"date":1534957200000,"team1":{"name":"HAVU","id":7865},"team2":{"name":"WASD","id":7668},"format":"bo3","event":{"name":"Thunderpick Invitational #3","id":3946},"stars":0,"live":false},{"id":2325782,"date":1534960800000,"team1":{"name":"DreamEaters","id":8305},"team2":{"name":"HAVU","id":7865},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325617,"date":1534978800000,"team1":{"name":"Wild","id":8895},"team2":{"name":"Coscu Army","id":8871},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325618,"date":1534983000000,"team1":{"name":"C4","id":8051},"team2":{"name":"YeaH!","id":8728},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325619,"date":1534987200000,"team1":{"name":"Furious","id":6799},"team2":{"name":"FURIA Inagame","id":8885},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325653,"date":1534987800000,"team1":{"name":"CERUS","id":7621},"team2":{"name":"Guerrilla Tactics","id":7882},"format":"bo3","event":{"name":"ESEA Advanced Season 28 North America","id":3909},"stars":0,"live":false},{"id":2325723,"date":1534987800000,"team1":{"name":"YeaH!","id":8728},"team2":{"name":"Bull","id":9203},"format":"bo3","event":{"name":"CGL Professional Season 3","id":3901},"stars":0,"live":false},{"id":2325620,"date":1534991400000,"team1":{"name":"Isurus","id":7653},"team2":{"name":"DETONA","id":8902},"format":"bo1","event":{"name":"Alienware Liga Pro Gamers Club - AUG/18","id":3967},"stars":0,"live":false},{"id":2325722,"date":1535072400000,"team1":{"name":"FURIA Inagame","id":8885},"team2":{"name":"W7M","id":8574},"format":"bo5","event":{"name":"GG.BET Summer Brazil","id":3955},"stars":0,"live":false},{"id":2325736,"date":1535077800000,"team1":{"name":"TYLOO","id":4863},"team2":{"name":"Kinguin","id":6134},"format":"bo3","event":{"name":"ZOTAC Cup Masters 2018 Grand Finals","id":3603},"stars":1,"live":false},{"id":2325737,"date":1535091000000,"team1":{"name":"OpTic","id":6615},"team2":{"name":"Ghost","id":7801},"format":"bo3","event":{"name":"ZOTAC Cup Masters 2018 Grand Finals","id":3603},"stars":1,"live":false},{"id":2325738,"date":1535104500000,"team1":{"name":"Virtus.pro","id":5378},"team2":{"name":"MVP PK","id":7354},"format":"bo3","event":{"name":"ZOTAC Cup Masters 2018 Grand Finals","id":3603},"stars":1,"live":false},{"id":2325739,"date":1535117700000,"team1":{"name":"MIBR","id":9215},"team2":{"name":"Flash","id":7660},"format":"bo3","event":{"name":"ZOTAC Cup Masters 2018 Grand Finals","id":3603},"stars":1,"live":false},{"id":2325779,"date":1535468400000,"team1":{"name":"forZe","id":8135},"team2":{"name":"Quantum Bellator Fire","id":7367},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325778,"date":1535472000000,"team1":{"name":"3DMAX","id":4914},"team2":{"name":"Jordan's Money Crew","id":9358},"format":"bo1","event":{"name":"Kalashnikov CUP","id":3977},"stars":0,"live":false},{"id":2325740,"date":1535536800000,"team1":{"name":"Astralis","id":6665},"team2":{"name":"Grayhound","id":8008},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":1,"live":false},{"id":2325741,"date":1535536800000,"team1":{"name":"Natus Vincere","id":4608},"team2":{"name":"Ghost","id":7801},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":1,"live":false},{"id":2325742,"date":1535542200000,"team1":{"name":"North","id":7533},"team2":{"name":"TYLOO","id":4863},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":2,"live":false},{"id":2325743,"date":1535542200000,"team1":{"name":"NRG","id":6673},"team2":{"name":"NiP","id":4411},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":2,"live":false},{"id":2325744,"date":1535547600000,"team1":{"name":"FaZe","id":6667},"team2":{"name":"Heroic","id":7175},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":1,"live":false},{"id":2325745,"date":1535547600000,"team1":{"name":"mousesports","id":4494},"team2":{"name":"Gambit","id":6651},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":2,"live":false},{"id":2325746,"date":1535553000000,"team1":{"name":"fnatic","id":4991},"team2":{"name":"OpTic","id":6615},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":2,"live":false},{"id":2325747,"date":1535553000000,"team1":{"name":"MIBR","id":9215},"team2":{"name":"HellRaisers","id":5310},"format":"bo1","event":{"name":"DreamHack Masters Stockholm 2018","id":3389},"stars":2,"live":false},{"id":2325748,"date":1535558400000,"title":"Group A elimination match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325749,"date":1535558400000,"title":"Group B elimination match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325750,"date":1535616000000,"title":"Group C elimination match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325752,"date":1535616000000,"title":"Group A winners' match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325751,"date":1535626800000,"title":"Group D elimination match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325753,"date":1535626800000,"title":"Group B winners' match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325754,"date":1535637600000,"title":"Group C winners' match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325755,"date":1535648400000,"title":"Group D winners' match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325756,"date":1535702400000,"title":"Group A decider match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325757,"date":1535713200000,"title":"Group B decider match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325758,"date":1535724000000,"title":"Group C decider match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325759,"date":1535734800000,"title":"Group D decider match - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325760,"date":1535788800000,"title":"Quarter-final #1 - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325761,"date":1535799600000,"title":"Quarter-final #2 - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325762,"date":1535810400000,"title":"Quarter-final #3 - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325763,"date":1535821200000,"title":"Quarter-final #4 - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325764,"date":1535878800000,"title":"Semi-final #1 - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325765,"date":1535889600000,"title":"Semi-final #2 - DreamHack Masters Stockholm 2018","stars":0,"live":false},{"id":2325766,"date":1535904000000,"title":"Grand final - DreamHack Masters Stockholm 2018","stars":0,"live":false}]`

**Comment by lucaspalencia** [comment]
@DestyKarpov does this work for other methods too?

**Comment by DestyKarpov** [comment]
@lucaspalencia Yeah, because the **fetchPage** function is used absolutely for all methods.
